components/async-container/async-container-mixin.js

```diff
@@ -0,0 +1,78 @@
+
+export const asyncStates = {
+	initial: 'initial',
+	pending: 'pending',
+	success: 'success',
+	failure: 'failure'
+};
+
+export const AsyncContainerMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			asyncState: { type: String }
+		};
+	}
+
+	constructor() {
+		super();
+		this._initializeAsyncState();
+		if (!this.asyncContainer) {
+			this.addEventListener('pending-state', this._handleAsyncItemState.bind(this));
+		}
+	}
+
+	firstUpdated() {
+		if (!this.asyncContainer) return;
+		const container = this.asyncContainer();
```

Comment:
If the consuming component defines an `asyncContainer` provider, then wire up to it instead of the host.

Initially I did not have this logic, and when I tried to use the mixin with `d2l-dialog`, I noticed that the initial `asyncState` was always `success`, even though the content was delayed by several seconds. The reason is - `d2l-dialog` uses `d2l-button-icon` (for abort in top-right of header), which uses `d2l-icon`, which uses `runAsync` directive. 

More generally, there can be various async items contained within the consumer of the mixin, and some async items may complete successfully before others even begin, resulting in a final mixin `asyncState` prematurely. So, `asyncContainer` enables the consuming element to limit the scope to listen for `pending-state`.

An alternative approach that I have been contemplating is to create some type of `AsyncObserver` instead of `AsyncContainerMixin`, and let consumers wire-up that way.  The downside is that it would not provide the handy `asyncState` property that consumers would generally use in their life cycle methods. The upside is that it perhaps gives them more control, and they may be able to use it in non Lit-element contexts.

Reply:
Hmm, I would not have expected `d2l-icon` to fall into this category of async things that observers would care about. It's definitely async, but it's fast enough that nobody should be waiting for it. So maybe it's just a matter of having them use different event names so they don't overlap?

Reply:
Yeah possibly.  We could also update the directive to accept options, one of which could be whether to dispatch the event.  🤔

Reply:
If I update the `runAsync` directive to accept an option for dispatching the event, what do you think the default should be?

Reply:
I'm not sure. Would the events just having different names "solve" the problem?

Reply:
So far I am thinking of three ways the event could be dispatched:

1. `runAsync` (dispatches `pending-state` per Justin's "Pending State Protocol")
2. `CachedTask` ( dispatches `pending-state`, also per Justin's "Pending State Protocol")
3. by async component developers manually dispatching the event (we could provide helper)

Any component that uses `runAsync` or `CachedTask` (if we chose to pull this in... from Justin's DevFest demo) could be a component we want to dispatch `pending-state`.  

In the `d2l-icons` case, it seems that we may not want the event dispatched, and in such case we don't really need any event at all.  We could update the directive to allow the consumer to override the event name, but I think Justin's idea is that this `pending-state` event would become something more commonly used across the community (although I'm mostly just concerned about what we decide for ourselves). 

I'm not really sure how the async component developer would choose whether to have the event dispatched or not, or if the consumer is allowed to decide the event name, how they choose when to use `pending-state` or something specific to their component.  🤔

Reply:
In general I agree that an individual component wouldn't decide whether to dispatch an event or not... it's up to the parent container as to whether it wants to handle async children. BUT, in icon's case, it's just doing a dynamic import of the icon SVG so it knows that it's going to be super fast AND it's also "handling" the pending-ness itself by reserving a blank properly sized area for the icon to load into. It knows that no parent should care about its pending state... it actually considers itself not async really at all.

Reply:
So in that case, if it's probably sufficient to provide the ability for `d2l-icon` to opt out of triggering the event?  i.e. no need to dispatch a different event.

Reply:
I think so yes.